R/module_bookmark_manager.R

```diff
@@ -0,0 +1,157 @@
+    bookmark_history <- reactiveVal({
+      list()
+    })
+
+    app_session$onBookmark(function(state) {
+      # Add current filter state to bookmark.
+      snapshot <- as.list(slices_global(), recursive = TRUE)
+      attr(snapshot, "mapping") <- matrix_to_mapping(mapping_matrix())
+      state$values$filter_state_on_bookmark <- snapshot
+      # Add snapshot history and bookmark history to bookmark.
+      state$values$snapshot_history <- snapshot_history()   # isolate this?
+      state$values$bookmark_history <- bookmark_history()   # isolate this?
+    })
+    app_session$onRestored(function(state) {
+      # Restore filter state.
+      snapshot <- state$values$filter_state_on_bookmark
+      snapshot_state <- as.teal_slices(snapshot)
+      mapping_unfolded <- unfold_mapping(attr(snapshot_state, "mapping"), names(filtered_data_list))
+      mapply(
+        function(filtered_data, filter_ids) {
+          filtered_data$clear_filter_states(force = TRUE)
+          slices <- Filter(function(x) x$id %in% filter_ids, snapshot_state)
+          filtered_data$set_filter_state(slices)
+        },
+        filtered_data = filtered_data_list,
+        filter_ids = mapping_unfolded
+      )
+      slices_global(snapshot_state)
+      # Restore snapshot history and bookmark history.
+      snapshot_history(state$values$snapshot_history)
+      bookmark_history(state$values$bookmark_history)
+    })
+
+    app_session$onBookmarked(function(url) {
+      bookmark_name <- trimws(input$bookmark_name)
+      if (identical(bookmark_name, "")) {
+        showNotification(
+          "Please name the bookmark.",
+          type = "message"
+        )
+        updateTextInput(inputId = "bookmark_name", value = "", placeholder = "Meaningful, unique name")
+        unlink(strsplit(url, "_state_id_=")[[1L]][[2L]], recursive = TRUE, force = TRUE, expand = FALSE)
+      } else if (is.element(make.names(bookmark_name), make.names(names(bookmark_history())))) {
+        showNotification(
+          "This name is in conflict with other bookmark names. Please choose a different one.",
+          type = "message"
+        )
+        updateTextInput(inputId = "bookmark_name", value = , placeholder = "Meaningful, unique name")
+        unlink(strsplit(url, "_state_id_=")[[1L]][[2L]], recursive = TRUE, force = TRUE, expand = FALSE)
+      } else {
+        # Add bookmark URL to bookmark history (with name).
+        bookmark_update <- c(bookmark_history(), list(url))
+        names(bookmark_update)[length(bookmark_update)] <- bookmark_name
+        bookmark_history(bookmark_update)
+
+        removeModal()
+      }
+    })
+
+    # Bookmark current input state - name bookmark.
+    observeEvent(input$bookmark_add, {
+      showModal(
+        modalDialog(
+          textInput(ns("bookmark_name"), "Name the bookmark", width = "100%", placeholder = "Meaningful, unique name"),
+          footer = tagList(
+            bookmarkButton("Accept", icon = icon("thumbs-up")),
```

Comment:
I tested [basic teal-app](https://github.com/insightsengineering/teal.gallery/blob/main/basic-teal/app.R) and button doesn't seem to work. Could you confirm?

Reply:
Replaced `bookmarkButton` with `actionButton`.